### PR TITLE
Made withAuthentication HOC

### LINKED ISSUE #181: When logged in, there is a flash of sign in page on first load
## Steps to reproduce

1) Sign in
2) Open a new tab, open boxwise

## Expected

A loading indicator is displayed, then I land on dashboard.

## What actually happened

The sign in page displayed for a second, then a loading indicator displayed, then I landed on the dashboard.

Probably related to #161

### PR DESCRIPTION
**Pull request by rubenspgcavalcante**
- Fixed auth. routing problem (no flashes anymore)
- Detached logic from within store and turned the authStateChange on a redux thunk
- Keep consistence between Routes (no mix of AuthedRoutes and Routes anymore)
- Keep user on the requested page if he/she is logged

Possible follow up on this (keep the user on the requested route after do the login)

closes #181

### CODE REVIEW

File: src/actions/auth.js [hunk 1]
```diff
@@ -2,6 +2,8 @@ import firebase from "../firebase";
 import { handleError } from "../utils";
 import { fetchProfile } from "./profile";
 
+export const USER_NOT_LOGGED = "USER_NOT_LOGGED";
```
Comment (bfirsh) [inline src/actions/auth.js]: A nit: `USER_NOT_SIGNED_IN` to be consistent with the other actions?

Comment (rubenspgcavalcante) [reply src/actions/auth.js]: Ok! If done better and now all the actions are using the async action factories.
I was thinking in something better than the first state being `loading: null` but for now looks good :)

**Comment by rubenspgcavalcante** [comment]
I'll need review again @bfirsh 😬

File: src/actions/auth.js [hunk 2]
```diff
@@ -1,24 +1,14 @@
 import firebase from "../firebase";
 import { handleError } from "../utils";
 import { fetchProfile } from "./profile";
+import { asyncAction, editAction } from "commons/utils/action-creators";
 
-export const USER_NOT_LOGGED = "USER_NOT_LOGGED";
+export const USER_SIGN_IN = asyncAction("USER_SIGN_IN");
+export const USER_NOT_SIGN_IN = "USER_NOT_LOGGED";
```
Comment (bfirsh) [inline src/actions/auth.js]: String needs updating :)

Comment (rubenspgcavalcante) [reply src/actions/auth.js]: sorry, working in multiple projects at the same time does that XD

**Pull request merged**